crates/components/src/lib.rs

```diff
@@ -96,11 +97,151 @@ where
+                if ui.add(self.widget).changed() {
+                    changed = true;
+                };
+            })
+            .response;
+        if changed {
+            response.mark_changed();
+        }
+        response
+    }
+}
+
+pub struct EnumComboBox<'a, T> {
+    id_source: egui::Id,
+    reference: &'a mut T,
+}
+
+impl<'a, T> EnumComboBox<'a, T> {
+    /// Creates a combo box that can be used to change the variant of an enum that implements
+    /// `strum::IntoEnumIterator + ToString`.
+    pub fn new(id_source: impl std::hash::Hash, reference: &'a mut T) -> Self {
+        Self {
+            id_source: egui::Id::new(id_source),
+            reference,
+        }
+    }
+}
+
+impl<'a, T> egui::Widget for EnumComboBox<'a, T>
+where
+    T: strum::IntoEnumIterator + ToString,
+{
+    fn ui(self, ui: &mut egui::Ui) -> egui::Response {
+        let mut changed = false;
+        let mut response = egui::ComboBox::from_id_source(self.id_source)
+            .wrap(true)
+            .width(ui.available_width() - ui.spacing().item_spacing.x)
+            .selected_text(self.reference.to_string())
+            .show_ui(ui, |ui| {
+                for (i, variant) in T::iter().enumerate() {
+                    let mut frame = egui::Frame::none();
+                    if i % 2 != 0 {
+                        frame = frame.fill(ui.visuals().faint_bg_color);
+                    }
+                    frame.show(ui, |ui| {
+                        if ui
+                            .selectable_label(
+                                std::mem::discriminant(self.reference)
+                                    == std::mem::discriminant(&variant),
+                                variant.to_string(),
+                            )
+                            .clicked()
+                        {
+                            *self.reference = variant;
+                            changed = true;
+                        }
+                    });
+                }
+            })
+            .response;
+        if changed {
+            response.mark_changed();
+        }
+        response
+    }
+}
+
+pub struct OptionalIdComboBox<'a, F>
+where
+    F: Fn(usize) -> String,
+{
+    id_source: egui::Id,
+    reference: &'a mut Option<usize>,
+    len: usize,
+    formatter: F,
+}
+
+impl<'a, F> OptionalIdComboBox<'a, F>
+where
+    F: Fn(usize) -> String,
+{
+    /// Creates a combo box that can be used to change the ID of an `optional_id` field in the data
+    /// cache.
+    pub fn new(
+        id_source: impl std::hash::Hash,
+        reference: &'a mut Option<usize>,
+        len: usize,
+        formatter: F,
+    ) -> Self {
+        Self {
+            id_source: egui::Id::new(id_source),
+            reference,
+            len,
+            formatter,
+        }
+    }
+}
+
+impl<'a, F> egui::Widget for OptionalIdComboBox<'a, F>
+where
+    F: Fn(usize) -> String,
+{
+    fn ui(self, ui: &mut egui::Ui) -> egui::Response {
```

Comment:
I'm not sure how possible this would be, but it'd be awesome if there was a way to search for things (like you can in the switch and variable picker)

Reply:
Okay, I added a search box with fuzzy search. I wanted to highlight where in each entry the searches match, but egui doesn't seem to allow changing the style for only part of a text label without messing up the label's colors.